### DISCUSSION TITLE
Increasing number of GPUs does not improve training speed

### DISCUSSION THREAD

**Discussion by wigging** [microsoft/mattergen#184] [Q&A]
I trained MatterGen with the MP-20 data using 4 GPUs where each GPU is an A100 with 80 GB of memory. The training stopped in about 4 hours and 50 minutes when it reached 899 epochs which is the default value for `max_epochs` in the config file. I used the following command to run the training:
```text
mattergen-train data_module=mp_20 ~trainer.logger trainer.devices=4
```
Next, I ran the same training using 8 GPUs where each GPU is an A100 with 80 GB of memory. The training stopped in about 4 hours and 53 minutes when it reached 899 epochs. I used the command shown below to run the training:
```text
mattergen-train data_module=mp_20 ~trainer.logger trainer.devices=8
```
I expected the training time to be shorter with 8 GPUs compared to 4 GPUs but it was about the same. Is MatterGen limited by the amount of memory that it can use? Is there a config setting that I need to adjust to take advantage of the extra memory provided by multiple GPUs?

**Comment by danielzuegner** [comment]
Hi @wigging,

did you verify that during training indeed all 8 GPUs were used? For MatterGen we always use a total batch size of 512, so for 4 GPUs this would be a per-GPU batch size of 128 and for 8 GPUs it is 64. You can increase this to make better use of your GPUs' memory, but you might have to tune the learning rate for the different batch size.

**Reply by wigging** [reply]
I used nvidia-smi to check the GPU usage during the training and I could see that all 8 GPUs were being used. However, I noticed that 40-48 GB out of the available 80 GB was being utilized for each GPU. So what batch size and learning rate do you recommend that I use?

**Reply by danielzuegner** [reply]
You can try doubling the per-GPU batch size for the 8-GPU run to 128. I'd recommend you also adjust the [learning rate](https://github.com/microsoft/mattergen/blob/main/mattergen/conf/lightning_module/default.yaml#L3) and [min learning rate](https://github.com/microsoft/mattergen/blob/main/mattergen/conf/lightning_module/default.yaml#L11). We're outside of what we were doing in the paper, but I think scaling these by a factor of 1.5 could be a good starting point.

**Reply by wigging** [reply]
I'm not sure how to define these in the command line argument but is this what you are suggesting:
```text
mattergen-train \
  data_module=mp_20 \
  ~trainer.logger \
  trainer.devices=8 \
  data_module.mp_20.batch_size=128 \
  lightning_module.optimizer_partial.lr=1.5e-4 \
  lightning_module.scheduler_partials.scheduler.min_lr=1.5e-6
```

**Reply by wigging** [reply]
When you say "double the batch size" what exactly are you referring to?

In `alex_mp_20.yaml` I see the following:

```yaml
batch_size:
  # total batch size of 512, adjust for number of devices, nodes, and gradient accumulation
  train:  ${eval:'(512 // ${trainer.accumulate_grad_batches}) // (${trainer.devices} * ${trainer.num_nodes})'}
  val: ${eval:'(512 // ${trainer.accumulate_grad_batches}) // (${trainer.devices} * ${trainer.num_nodes})'}
```

and in `mp_20.yaml` I see the following:

```yaml
batch_size:
  # total batch size of 512, adjust for number of devices, nodes, and gradient accumulation
  train:  ${eval:'(512 // ${trainer.accumulate_grad_batches}) // (${trainer.devices} * ${trainer.num_nodes})'}
  val: ${eval:'(64 // (${trainer.devices} * ${trainer.num_nodes}))'}
  test: ${eval:'(64 // (${trainer.devices} * ${trainer.num_nodes}))'}
```

Are you saying I should define the config settings as follows when using 8 GPUs:

```yaml
batch_size:
  train: 128
  val: 64
  test: 64
```

**Reply by wigging** [reply]
Does `trainer.accumulate_grad_batches` do anything other than act as a variable to calculate the batch size?

**Reply by danielzuegner** [reply]
Hi,

I meant increasing `512` in the training batch size `${eval:...}` block to `1024`. This is the effective batch size after accounting for multiple GPUs, multiple nodes, and gradient accumulation. `trainer.accumulate_grad_batches` will average the gradients over that many batches before taking an optimizer step, hence increasing the effective batch size by that number (see [documentation](https://lightning.ai/docs/pytorch/stable/common/trainer.html#accumulate-grad-batches)).

**Reply by wigging** [reply]
But why calculate the train value with `train: ${eval: ...}` when you can explicitly define it with `train: 128`? Also, do I need to adjust the values for `val` and `test` or are they not used for training?

**Reply by danielzuegner** [reply]
The thing is that we need to jointly tune the effective batch size with the other training hyperparameters such as learning rate. If we change the effective batch size without adjusting the learning rate accordingly, the training behavior will be different. Thus, we fix the effective batch size and adjust the per-GPU batch size based on the number of available GPUs and the gradient accumulation.

**Reply by wigging** [reply]
Ok, let us make sure I understand your suggestions before I run the training again.

Change `mattergen/conf/data_module/mp_20.yaml` to the following:

```yaml
batch_size:
  train: ${eval:'(1024 // ${trainer.accumulate_grad_batches}) // (${trainer.devices} * ${trainer.num_nodes})'}
```

Change `mattergen/conf/lightning_module/default.yaml` to the following:

```yaml
optimizer_partial:
  lr: 1.5e-4

scheduler_partials:
  - scheduler:
      min_lr: 1.5e-6
```

Run the training with this command:

```text
mattergen-train data_module=mp_20 ~trainer.logger trainer.devices=8
```

Is all of this correct?

**Reply by danielzuegner** [reply]
Yes, that looks right.

**Reply by wigging** [reply]
The total training time with MP-20 was 2 hours 35 minutes using the config settings above. This was with 8 H100s SXM5 with 80 GB of memory for each GPU. So these settings definitely improved the training time compared to my previous attempt.

**Reply by wigging** [reply]
One more thing... If I increased the number of GPUs again to something like 16 GPUs then what would I use for batch size and learning rates? Would I increase the batch size by a factor of 2 again and scale the learning rates by 1.5 once more? So for 16 GPUs the batch size would be 2048 and `lr` would be 2.25e-4 and `min_lr` is 2.25e-6.
